api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,629 @@
+		// https://github.com/aws/aws-sdk-go/blob/ee1f179877b2daf2aaabf71fa900773bf8842253/service/ecs/api.go#L5420
+		// (aws-sdk-go version 1.12.19, as stated in layer0/Gopkg.toml)
+		for i := 0; i < len(serviceARNs); i += 10 {
+			end := i + 10
+
+			if end > len(serviceARNs) {
+				end = len(serviceARNs)
+			}
+
+			describeServicesInput := &ecs.DescribeServicesInput{}
+			describeServicesInput.SetCluster(clusterName)
+			describeServicesInput.SetServices(serviceARNs[i:end])
+
+			output, err := e.Client.ECS.DescribeServices(describeServicesInput)
+			if err != nil {
+				return nil, err
+			}
+
+			services = append(services, output.Services...)
+		}
+	}
+
+	resourceConsumers := []models.ResourceConsumer{}
+
+	for _, service := range services {
+		deployIDCopies := map[string]int64{}
+		for _, d := range service.Deployments {
+			desiredCount := aws.Int64Value(d.DesiredCount)
+			runningCount := aws.Int64Value(d.RunningCount)
+			pendingCount := aws.Int64Value(d.PendingCount)
+			if numPending := desiredCount - (runningCount + pendingCount); numPending > 0 {
+				deployIDCopies[aws.StringValue(d.Id)] = numPending
+			}
+		}
+
+		if len(deployIDCopies) == 0 {
+			continue
+		}
+
+		// iterate through deploys
+		for deployID := range deployIDCopies {
+			c, err := e.getContainerResourceFromDeploy(deployID)
+			if err != nil {
+				return nil, err
+			}
+
+			resourceConsumers = append(resourceConsumers, c...)
+		}
+	}
+
+	// GET PENDING TASK RESOURCE CONSUMERS IN ECS
+	taskSummaries, err := e.TaskProvider.List()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, summary := range taskSummaries {
+		if summary.EnvironmentID == clusterName {
+
+			task, err := e.TaskProvider.Read(summary.TaskID)
+			if err != nil {
+				return nil, err
+			}
+
+			taskResourceConsumers, err := e.getContainerResourceFromDeploy(task.DeployID)
+			if err != nil {
+				return nil, err
+			}
+
+			resourceConsumers = append(resourceConsumers, taskResourceConsumers...)
+		}
+	}
+
+	// GET PENDING TASK RESOURCE CONSUMERS IN JOBS
+	jobs, err := e.JobStore.SelectAll()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, job := range jobs {
+		if job.Type == models.CreateDeployJob {
+			if job.Status == models.PendingJobStatus || job.Status == models.InProgressJobStatus {
+				var req models.CreateTaskRequest
+				if err := json.Unmarshal([]byte(job.Request), &req); err != nil {
+					return nil, err
+				}
+
+				if req.EnvironmentID == clusterName {
+					// note that this isn't exact if the job has started some, but not all of the tasks
+					taskResourceConsumers, err := e.getContainerResourceFromDeploy(req.DeployID)
+					if err != nil {
+						return nil, err
+					}
+
+					resourceConsumers = append(resourceConsumers, taskResourceConsumers...)
+				}
+			}
+		}
+	}
+
+	return resourceConsumers, nil
+}
+
+func (e *EnvironmentScaler) calculateNewProvider(clusterName string) (*models.ResourceProvider, error) {
+	input := &autoscaling.CreateAutoScalingGroupInput{}
```

Comment:
Why are we creating an AutoScalingGroup here? The whole purpose seems to be to get the InstanceType from a launch config. Can we not get this from environment.size?

1. Create ASG
2. Describe the launch config for the newly created group

Reply:
We could remove #L378-399 and then just get available memory from env size, using tag store right?